File: ModernPerfectSecrecy.py
import os
import sys
import hashlib

SIZE_OF_BLOCK = 512
SIZE_OF_INTEGER = 8
SIZE_OF_BLOCK_FOR_FILE = pow(2,SIZE_OF_INTEGER)* SIZE_OF_INTEGER

def generateSessionKey(hashedSecretKey, blockIndex, *filePaths):
	filechunks = []

	for index,filePath in enumerate(filePaths):
		f = file(filePath,"rb")
		f.seek(blockIndex*SIZE_OF_BLOCK_FOR_FILE/8)
		filechunks.append(list(chunks(f.read(SIZE_OF_BLOCK_FOR_FILE/8).encode('hex'),2)))

	secretKeyChunks = chunks(hashedSecretKey,2)
	orderMapping = [None]*(pow(2,SIZE_OF_INTEGER))
	MODULUS = pow(2,SIZE_OF_INTEGER)

	sessionKey = []
	# 8 bit integer in secret key
	for index in secretKeyChunks:
		value = int(index,16)
		# collision resolution
		while orderMapping[value] == 1:
			value = (value + 1) % MODULUS
		orderMapping[value]=1

		index_of_block = value
		# file chaining
		for individualFile in filechunks:
			index_of_block = int(individualFile[index_of_block],16)
		sessionKey.append(("%.2x"%index_of_block))
	sessionKey = ''.join(sessionKey)

	return sessionKey

def chunks(l, n):
    for i in xrange(0, len(l), n):
        yield l[i:i+n]

def protectSecretKey(secretKey, nonce1, nonce2):
	print "PreSharedKey(hex) : " + secretKey
	secretKeyInt = int(secretKey,16)

	print "Nonce1 : " + nonce1
	nonce1Int = int(nonce1,16)

	print "Nonce2 : " + nonce2 
	nonce2Int = int(nonce2,16)

	sessionSecretKey = secretKeyInt ^ nonce1Int ^ nonce2Int
	sessionSecretKey = hashlib.sha512('%x'%sessionSecretKey).hexdigest()
	print "Session Secret Key : " + sessionSecretKey

	return sessionSecretKey

def encrypt(secretKey, nonce1, nonce2, plaintextFile, *filePaths):
	print "Secret Key: %s"%secretKey
	sessionSecretKey = protectSecretKey(secretKey,nonce1,nonce2)
	print "-" * 80
	plaintext = file(plaintextFile, "rb").read()
	plaintextHex = plaintext.encode('hex')
	plaintextChunks = chunks(plaintextHex,SIZE_OF_BLOCK/8*2)

	cipherText = ""
	for index,plainChunk in enumerate(plaintextChunks):
		sessionKeyForChunk = generateSessionKey(sessionSecretKey, index, *filePaths)
		paddedPlainChunk =  plainChunk.ljust(SIZE_OF_BLOCK/8*2,'0')
		cipherTextForChunk = "%.128x"% (int(paddedPlainChunk,16) ^ int(sessionKeyForChunk,16))
		cipherText += cipherTextForChunk
	return cipherText

def encryptCBC(secretKey, nonce1, nonce2, plaintextFile, *filePaths):
	print "Secret Key: %s"%secretKey
	sessionSecretKey = protectSecretKey(secretKey,nonce1,nonce2)
	print "-" * 80
	plaintext = file(plaintextFile, "rb").read()
	plaintextHex = plaintext.encode('hex')
	plaintextChunks = chunks(plaintextHex,SIZE_OF_BLOCK/8*2)

	cipherText = ""
	previousCipherText = 0
	for index,plainChunk in enumerate(plaintextChunks):
		sessionKeyForChunk = generateSessionKey(sessionSecretKey, index, *filePaths)
		paddedPlainChunk =  plainChunk.ljust(SIZE_OF_BLOCK/8*2,'0')
		cipherTextForChunk = "%.128x"% (previousCipherText^int(paddedPlainChunk,16) ^ int(sessionKeyForChunk,16))
		previousCipherText = int(cipherTextForChunk,16)
		cipherText += cipherTextForChunk
	return cipherText

def decryptCBC(secretKey, nonce1, nonce2, cipherTextFile, *filePaths):
	print "Secret Key: %s"%secretKey
	sessionSecretKey = protectSecretKey(secretKey,nonce1,nonce2)
	print "-" * 80
	cipherText = file(cipherTextFile, "rb").read()
	cipherTextHex = cipherText.encode('hex')
	cipherTextChunks = chunks(cipherTextHex,SIZE_OF_BLOCK/8*2)

	plainText = ""
	previousCipherText = 0
	for index,cipherChunk in enumerate(cipherTextChunks):
		sessionKeyForChunk = generateSessionKey(sessionSecretKey, index, *filePaths)
		plainTextForChunk = "%.128x"% (previousCipherText^int(cipherChunk,16) ^ int(sessionKeyForChunk,16))
		previousCipherText = int(cipherChunk,16)
		plainText += plainTextForChunk
	return plainText

#############################################################################################

def sampleECB():
	secretKey = file("secretkey","rb").read()
	nonce1 = file("nonce1","rb").read()
	nonce2 = file("nonce2","rb").read()
	cipherText= encrypt(secretKey, nonce1, nonce2, "samplePlainText.txt", "file1.m4a","file2.avi")
	print "CipherText: %s"%cipherText.decode('hex')
	temp = file("/tmp/input","wb")
	temp.write(cipherText.decode('hex'))
	temp.close()
	print "-" * 80
	print "PlainText: %s"%encrypt(secretKey, nonce1, nonce2, "/tmp/input", "file1.m4a","file2.avi").decode('hex')	

def sampleCBC():
	secretKey = file("secretkey","rb").read()
	nonce1 = file("nonce1","rb").read()
	nonce2 = file("nonce2","rb").read()
	cipherText= encryptCBC(secretKey, nonce1, nonce2, "samplePlainText.txt", "file1.m4a","file2.avi")
	print "CipherText: %s"%cipherText.decode('hex')
	temp = file("/tmp/input","wb")
	temp.write(cipherText.decode('hex'))
	temp.close()
	print "-" * 80
	print "PlainText: %s"%decryptCBC(secretKey, nonce1, nonce2, "/tmp/input", "file1.m4a","file2.avi").decode('hex')	

def sampleBMP():
	secretKey = file("secretkey","rb").read()
	nonce1 = file("nonce1","rb").read()
	nonce2 = file("nonce2","rb").read()
	cipherText= encrypt(secretKey, nonce1, nonce2, "sample.bmp", "file1.m4a","file2.avi")
	temp = file("./encryptedSampleECB.bmp","wb")
	temp.write(cipherText.decode('hex'))
	temp.close()

	cipherText= encryptCBC(secretKey, nonce1, nonce2, "sample.bmp", "file1.m4a","file2.avi")
	temp = file("./encryptedSampleCBC.bmp","wb")
	temp.write(cipherText.decode('hex'))
	temp.close()


def main(argv):
	sampleECB()
	sampleBMP()

if __name__ == "__main__":
	main(sys.argv)

File: ModernPerfectSecrecy.c
#include <stdio.h>
#include <stdarg.h>
#include <math.h>
#include <stdlib.h>
#include <string.h>
#include <openssl/sha.h>

#define _FILE_OFFSET_BITS 64

int SIZE_OF_BLOCK; //bits
int SIZE_OF_INTEGER; //bits
int SIZE_OF_BLOCK_FOR_FILE; //bits
int BLOCK_FOR_FILE_BITS; //should be bytes not bits
int NUMBER_OF_BLOCKS;
int LENGTH_OF_BLOCK; //64 bytes, 128 characters
int LENGTH_OF_BLOCK_IN_BYTES; // 64
int MODULUS;

struct node {
  char data[65];
  struct node *next;
};

unsigned char** generateBlockKey(char* sessionKey,int numberOfChunks,int numOfFile, char** listOfFiles) {

	//We only require numberOfChunks from each file
	char*** filechunks = malloc(sizeof(char**) * numOfFile);
	int i;
	for (i = 0; i<numOfFile;++i) {
		if ( (filechunks[i] = malloc(sizeof(char*)*numberOfChunks)) == NULL) {

		}
		int x;
		for (x=0;x<numberOfChunks;++x) {
			if ( (filechunks[i][x] = malloc(sizeof(char)*(NUMBER_OF_BLOCKS+1))) == NULL) {

			}
		}
	}

	for (i = 0;i < numOfFile; ++i) {
		char* filePath = listOfFiles[i];
		char* mode = "rb";
		FILE* inputFile = fopen(filePath, mode);

		if (inputFile == NULL) {
			fprintf(stderr, "Can't open input file provided!\n");
  			exit(1);
		}
		int num = 0;
		while (!feof(inputFile) && num < numberOfChunks) {
			fread(filechunks[i][num], 1, NUMBER_OF_BLOCKS, inputFile);
			num++;
		}
		fclose(inputFile);
	}

	int lengthOfKey = LENGTH_OF_BLOCK;
	int lengthOfBlockKey = LENGTH_OF_BLOCK_IN_BYTES;
	int secretKey_Which[lengthOfBlockKey]; // 2 char per integer, so half of length , 64 bytes
	int mapping[NUMBER_OF_BLOCKS];
	memset(mapping,0,sizeof(mapping));
	for (i = 0; i < LENGTH_OF_BLOCK_IN_BYTES; ++i) {
		unsigned char value = sessionKey[i];

		while(mapping[value]==1) {
			value = (value+1) % MODULUS;
		}
		mapping[value]=1;
		secretKey_Which[i]=value;
	}

	unsigned char ** blockKeys = malloc( sizeof(char*) * (numberOfChunks));
	for (i=0; i< numberOfChunks;++i) {
		if ( (blockKeys[i] = malloc(lengthOfBlockKey+1)) == NULL) {

		}
	}
	for (i=0; i<numberOfChunks; ++i) {
		unsigned char blockKey[lengthOfBlockKey];
		memset(blockKey,0,sizeof(blockKey));
		int index;
		for (index=0;index<lengthOfBlockKey;++index) {
		 	int indexOfBlock = secretKey_Which[index];
			// file chainings
			int fileIndex;
			for (fileIndex = 0; fileIndex < numOfFile;++fileIndex) {
				unsigned char next_index = filechunks[fileIndex][i][indexOfBlock];
				indexOfBlock = next_index;
			}
			blockKeys[i][index] = indexOfBlock;
		}
	}

	return blockKeys;
}

char* generateSessionSecretKey(char* secretKey,char* nonce1,char* nonce2) {
	long long secretKey_int = (int) strtoll(secretKey,NULL,16);
	long long nonce1_int = (int)strtoll(nonce1,NULL,16);
	long long nonce2_int = (int)strtoll(nonce2,NULL,16);

	secretKey_int = secretKey_int ^ nonce1_int ^ nonce2_int;
	char* sessionKey = malloc(LENGTH_OF_BLOCK_IN_BYTES+1);
	snprintf(sessionKey, LENGTH_OF_BLOCK_IN_BYTES+1,"%64llX", secretKey_int);

	SHA512(sessionKey, sizeof(sessionKey)-1, sessionKey); 

	return sessionKey;
}

unsigned char* encrypt(char* secretKey, char* nonce1, char* nonce2, char* plaintextFile, int numOfFile, char** listOfFiles) {
	char* sessionKey = generateSessionSecretKey(secretKey, nonce1, nonce2);

	char* mode = "rb";
	FILE* inputFile = fopen(plaintextFile, mode);

	if (inputFile == NULL) {
		fprintf(stderr, "Can't open plaintext file provided! %s\n",plaintextFile);
		exit(1);
	}

	struct node *root;
	root = (struct node *) malloc( sizeof(struct node));

	struct node *cur;
	cur = root;
	fseek(inputFile, 0, SEEK_END);
    long f_size = ftell(inputFile);
    fseek(inputFile, 0, SEEK_SET);

	int numberOfChunks = ceil((double)f_size/LENGTH_OF_BLOCK_IN_BYTES);
	int i = 0;
	while (!feof(inputFile) && i < numberOfChunks) {
		i++;
		fread(cur->data, 1, LENGTH_OF_BLOCK_IN_BYTES, inputFile);
		struct node *next = (struct node *) malloc( sizeof(struct node));
		cur->next=next;
		cur=next;
	}
	fclose(inputFile);

	unsigned char**  blockKeys = generateBlockKey(sessionKey,numberOfChunks,numOfFile, listOfFiles);

	unsigned char* cipherText= malloc(numberOfChunks*LENGTH_OF_BLOCK_IN_BYTES+1);
	cur = root;

	i = 0;
	int x= 0;
	while(cur->next != NULL) {
		unsigned char* blockKey = blockKeys[i];
		unsigned char* data = (unsigned char*)cur->data;

		unsigned char result[LENGTH_OF_BLOCK_IN_BYTES];

		int index;
		for (index = 0;index < LENGTH_OF_BLOCK_IN_BYTES; ++index) {
			x++;
			int offset = i * LENGTH_OF_BLOCK_IN_BYTES;
			cipherText[offset+index] = blockKey[index] ^ data[index];
		}
		cur = cur->next;
		i++;
	}

	fwrite(cipherText, 1, f_size, stdout);

	return cipherText;
}

char* readFile(char* filePath) {
	FILE *f = fopen(filePath, "rb");
	if (f == NULL) {
		fprintf(stderr, "Can't open file provided! %s\n",filePath);
		exit(1);
	}

	fseek(f, 0, SEEK_END);
	long fsize = ftell(f);
	fseek(f, 0, SEEK_SET);

	char *string = malloc(fsize + 1);
	fread(string, fsize, 1, f);
	fclose(f);

	string[fsize] = 0;

	return string;
}

int main(int argc, char* argv[]) {
	SIZE_OF_BLOCK = 512; //bits
	SIZE_OF_INTEGER = 8; //bits
	NUMBER_OF_BLOCKS = pow(2,SIZE_OF_INTEGER);
	MODULUS = NUMBER_OF_BLOCKS;
	SIZE_OF_BLOCK_FOR_FILE = NUMBER_OF_BLOCKS << 3; // number of bits
	LENGTH_OF_BLOCK = SIZE_OF_BLOCK >> 2; //64 bytes, 128 characters
	LENGTH_OF_BLOCK_IN_BYTES = LENGTH_OF_BLOCK>>1;

	if (argc<5) {
		printf("Usage: %s secretKeyPath nonce1Path nonce2Path plaintextfilepath file1 file2 ...", argv[0]);
	}

	char* secretKeyPath = argv[1];
	char* nonce1Path = argv[2];
	char* nonce2Path = argv[3];
	char* plaintextPath = argv[4];

	char* secretKey = readFile(secretKeyPath);
	char* nonce1 = readFile(nonce1Path);
	char* nonce2 = readFile(nonce2Path);

	char** listOfFiles = malloc(sizeof(char*) * (argc-5));
	int i;
	for (i=0;i<argc-5;++i) {
		listOfFiles[i] = argv[i+5];
	}

	encrypt(secretKey,nonce1,nonce2,plaintextPath,(argc-5),listOfFiles);
	return 0;
}

